xarray/core/groupby.py

```diff
@@ -294,92 +311,336 @@ def _apply_loffset(
+
+        return codes, group_indices, unique_coord, full_index
+
+
+class Grouper(ABC):
+    pass
+
+
+@dataclass
+class UniqueGrouper(Grouper):
+    pass
+
+
+@dataclass
+class BinGrouper(Grouper):
+    bins: Any  # TODO: What is the typing?
+    cut_kwargs: Mapping = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        if duck_array_ops.isnull(self.bins).all():
+            raise ValueError("All bin edges are NaN.")
+
+
+@dataclass
+class TimeResampleGrouper(Grouper):
+    freq: str
+    closed: SideOptions | None
+    label: SideOptions | None
+    origin: str | DatetimeLike | None
+    offset: pd.Timedelta | datetime.timedelta | str | None
+    loffset: datetime.timedelta | str | None
+
+
+def _validate_groupby_squeeze(squeeze: bool) -> None:
+    # While we don't generally check the type of every arg, passing
+    # multiple dimensions as multiple arguments is common enough, and the
+    # consequences hidden enough (strings evaluate as true) to warrant
+    # checking here.
+    # A future version could make squeeze kwarg only, but would face
+    # backward-compat issues.
+    if not isinstance(squeeze, bool):
+        raise TypeError(f"`squeeze` must be True or False, but {squeeze} was supplied")
+
+
+def _resolve_group(obj: T_Xarray, group: T_Group | Hashable) -> T_Group:
+    from xarray.core.dataarray import DataArray
+
+    error_msg = (
+        "the group variable's length does not "
+        "match the length of this variable along its "
+        "dimensions"
+    )
+
+    newgroup: T_Group
+    if isinstance(group, DataArray):
+        try:
+            align(obj, group, join="exact", copy=False)
+        except ValueError:
+            raise ValueError(error_msg)
+
+        newgroup = group.copy(deep=False)
+        newgroup.name = group.name or "group"
+
+    elif isinstance(group, IndexVariable):
+        # This assumption is built in to _ensure_1d.
+        if group.ndim != 1:
+            raise ValueError(
+                "Grouping by multi-dimensional IndexVariables is not allowed."
+                "Convert to and pass a DataArray instead."
+            )
+        (group_dim,) = group.dims
+        if len(group) != obj.sizes[group_dim]:
+            raise ValueError(error_msg)
+
     else:
-        group_indices = np.arange(group.size)
-    codes = np.arange(group.size)
-    unique_coord = group
-    return unique_coord, group_indices, codes
+        if not hashable(group):
+            raise TypeError(
+                "`group` must be an xarray.DataArray or the "
+                "name of an xarray variable or dimension. "
+                f"Received {group!r} instead."
+            )
+        group = obj[group]
+        if group.name not in obj._indexes and group.name in obj.dims:
+            # DummyGroups should not appear on groupby results
+            newgroup = _DummyGroup(obj, group.name, group.coords)
+        else:
+            newgroup = group
+
+    if newgroup.size == 0:
```

Comment:
With xarray 2023.5.0 I seem to now get "UnboundLocalError: local variable 'newgroup' referenced before assignment" when using groupby with a IndexVariable object.

Reply:
Can you open a new issue please?

Reply:
See #7919